### PR TITLE
databroker: add support for querying the databroker

### PR DESCRIPTION
**Pull request by calebdoxsey**
## Summary
This adds a new method to the databroker to query it based on a string with an offset and limit.

**Checklist**:
- [ ] add related issues
- [ ] updated docs
- [x] updated unit tests
- [ ] updated UPGRADING.md
- [x] ready for review

### CODE REVIEW

File: pkg/storage/redis/redis.go [hunk 1]
```diff
@@ -256,6 +258,25 @@ func (db *DB) ClearDeleted(ctx context.Context, cutoff time.Time) {
 	}
 }
 
+// Query queries for records.
+func (db *DB) Query(ctx context.Context, query string, offset, limit int) (records []*databroker.Record, totalCount int, err error) {
+	query = strings.ToLower(query)
```
Comment (travisgroth) [inline pkg/storage/redis/redis.go:263]:
```suggestion
	_, span := trace.StartSpan(ctx, "databroker.redis.Query")
	defer span.End()
	
	query = strings.ToLower(query)
```

**Review by travisgroth — APPROVED** [review]
One nit on tracing but LGTM

**Comment by calebdoxsey** [comment]
@travisgroth I forgot about encryption. Since we store the data encrypted, it's not possible to implement searching on the backend. Instead I switched it to use the GetAll method and do the searching in the server gRPC method.

File: pkg/storage/storage_test.go [hunk 2]
```diff
@@ -47,6 +51,20 @@ func (m *mockBackend) ClearDeleted(ctx context.Context, cutoff time.Time) {
...
 }
+
+func TestMatchAny(t *testing.T) {
+	u := &user.User{Id: "id", Name: "name", Email: "email"}
+	data, _ := anypb.New(u)
+	assert.True(t, MatchAny(data, ""))
+	assert.True(t, MatchAny(data, "id"))
+	assert.True(t, MatchAny(data, "name"))
+	assert.True(t, MatchAny(data, "email"))
+	assert.False(t, MatchAny(data, "nope"))
```
Comment (desimone) [inline pkg/storage/storage_test.go:69]: _nope_

File: pkg/storage/storage.go [hunk 3]
```diff
@@ -38,3 +41,74 @@ type Backend interface {
...
+	matches := false
+
+	md := msg.Descriptor()
+	fds := md.Fields()
+	for i := 0; i < fds.Len(); i++ {
+		fd := fds.Get(i)
+		if !msg.Has(fd) {
+			continue
+		}
+		matches = matches || matchProtoValue(fd, msg.Get(fd), query)
```
Comment (desimone) [inline pkg/storage/storage.go:72]: Can we return early if a match was found?

File: internal/databroker/server.go [hunk 4]
```diff
@@ -220,6 +220,33 @@ func (srv *Server) GetAll(ctx context.Context, req *databroker.GetAllRequest) (*
 	}, nil
 }
 
+// Query queries for records.
+func (srv *Server) Query(ctx context.Context, req *databroker.QueryRequest) (*databroker.QueryResponse, error) {
+	_, span := trace.StartSpan(ctx, "databroker.grpc.Query")
```
[the remark this answers is not in the document]
Comment (calebdoxsey) [reply internal/databroker/server.go:226]: I removed db.Query.

**Pull request merged**